etcdserver/api/v3rpc/maintenance.go

```diff
@@ -128,11 +128,11 @@ func (ms *maintenanceServer) Snapshot(sr *pb.SnapshotRequest, srv pb.Maintenance
 }
 
 func (ms *maintenanceServer) Hash(ctx context.Context, r *pb.HashRequest) (*pb.HashResponse, error) {
```

Comment:
I went with @heyitsanthony suggestion. one issue with current full hash operation it ignores compaction. However, I can probably merge `Hash()` and `HashByRev` in to one function so both full hash and hashByRev will wait for compaction.

something like following
```go
func (s *store) HashByRev(rev int64) (hash uint32, currentRev int64, compactRev int64, err error) {
	for {
		s.mu.RLock()
		compactMainRev := s.compactMainRev
		currentRev = s.currentRev
		compactc := s.compactc
		s.mu.RUnlock()
		if rev > 0 && rev < compactMainRev {
			return 0, 0, compactMainRev, ErrCompacted
		} else if rev > 0 && rev > currentRev {
			return 0, currentRev, 0, ErrFutureRev
		}
		select {
		case <-compactc:
		case <-s.stopc:
			return 0, 0, 0, ErrClosed
		}
		s.mu.Lock()
		if s.compactc == compactc {
			defer s.mu.Unlock()
			break
		}
		s.mu.Unlock()
	}
	s.b.ForceCommit()

	if rev == 0 { // full hash if rev == 0
		hash, err = s.b.Hash(DefaultIgnores)
	} else {
		upper := revision{main: rev + 1}
		f := func(r []byte) bool {
			return upper.GreaterThan(bytesToRev(r))
		}

		hash, err = s.b.HashBucket(keyBucketName, f)
	}
	return hash, s.currentRev, s.compactMainRev, err
}
```